mgr/cephadm: revamp scheduling

liewegas opened the pull request:
- Reimplement HostAssignment to respect labels, all:true, etc.
- Update the orch CLI to take an array of strings for all placement specs

Note: I wonder if it would be better to take a *single* string for placement, and make it comma or ; separated if there are multiple items to include...

liewegas commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -1948,26 +1948,51 @@ def _apply_service(self, daemon_type, spec, create_func, config_func=None):
+        def _then_rm(so_far):
+            r = [trivial_result(t) for t in so_far]
+            r.extend(self._remove_daemon(rm_args))
+            return r
```
@sebastian-philipp not quite sure how to do this so that the result is a list of "added ..." and "removed ..." messages in the end...

liewegas commented:
@sebastian-philipp i ran out of time today and wasn't able to figure out how to do this.  ideally i'd like it to do all of the add items in parallel, then all of the delete items.  but even if it did them all serially, or adds and removes in parallel, i think would be good enough for now...

The code right now fails with ``AttributeError: 'function' object has no attribute 'has_result'``.

sebastian-philipp commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -1948,26 +1948,41 @@ def _apply_service(self, daemon_type, spec, create_func, config_func=None):
+                create_func, config_func))
+        if rm_args:
+            r.extend(self._remove_daemon(rm_args))
+        return async_map_completion(r)
```
Having those async_map_completions within the serve() thread is really questionable. Would probably make sense to extract the `map_completion` from the `async` part? 

would that help?

liewegas replied:
yeah. ideally we want the parallelism but maybe don't need the rest?

liewegas commented:
@sebastian-philipp I dropped the commit that changes apply_service so that we can get the HostAssignement changes in.  Added a commit to take a uniform placement strings array for all the CLI commands.

liewegas commented:
retest this please

sebastian-philipp commented:
https://tracker.ceph.com/issues/44167 needs to be considered as well.

liewegas commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -2428,65 +2429,132 @@ def __init__(self,
+
+        # we need to select a subset of the candidates
+
+        # prefer hosts that already have services
```
This isn't colocating--it's "choosing" the host where a daemon already exists.  So if you have count=3 for mds.foo and there are 3 MDSs on hosts a, b, and c, then NodePlcaement should return a,b,c and not d,e,f (which would *move* everything needlessly).

liewegas commented:
> https://tracker.ceph.com/issues/44167 needs to be considered as well.

This PR (now) handles just the NodeAssignment part.  The apply_service method needs to be updated to handle daemon adds/removes properly, but i dropped that patch since it didn't work yet and will probably get redone with the persistent spec approach anyway.

liewegas commented:
retest this please

liewegas commented:
green: http://pulpito.ceph.com/sage-2020-02-29_20:08:05-rados:cephadm-wip-sage3-testing-2020-02-29-1048-distro-basic-smithi/#

liewegas commented:
rebased on top of the spec apply PR, still green: http://pulpito.ceph.com/sage-2020-03-01_22:29:51-rados:cephadm-wip-sage3-testing-2020-03-01-1419-distro-basic-smithi/

liewegas commented:
retest this please

tchaikov commented:
- http://pulpito.ceph.com/kchai-2020-02-29_17:36:34-rados-wip-kefu-testing-2020-02-29-2106-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/

jschmid1 commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -2875,68 +2875,145 @@ class HostAssignment(object):
+            hosts = self.pick_candidates()
+
+        if not self.spec.placement.count:
+            logger.info('Labeled hosts: {}'.format(hosts))
```
Those are not necessarily 'labeled' hosts. The conditional's if branch from above assigns explicitly set hosts to `hosts` if `self.placement.hosts` is populated.

liewegas replied:
the first branch is only taken if count is true.  it's convoluted, i know.. i was struggling to make is clearer, but it's hard when all of the fields have slightly different meanings when used in combination with each other.  i think i covered all of the cases in the tests

jschmid1 replied:
yep, the tests certainly helped!
